Dockerfile

```diff
@@ -0,0 +1,15 @@
+FROM node:9.3.0
+
+RUN apt-get update && apt-get install -y rsync
+
+RUN mkdir -p /usr/src/app
+
+COPY . /usr/src/app/
+RUN chmod a+x /usr/src/app/wait-for-it.sh
+
+WORKDIR /usr/src/app
+
+RUN npm install
+RUN npm run build
+
+CMD [ "npm", "start" ]
```

Comment:
If we can rely on caching layers, rebuilds of the container will go faster using this ordering:
```
FROM node:9.11.1

RUN apt-get update && apt-get install -y rsync

RUN mkdir -p /usr/src/app
WORKDIR /usr/src/app

COPY package.json package-lock.json .
RUN npm install

COPY . /usr/src/app/
RUN chmod a+x /usr/src/app/wait-for-it.sh
RUN npm run build

CMD [ "npm", "start" ]
```

Reply:
so multiple COPY commands adds an additional layer. each directive in the Dockerfile is seen as Docker image layer on file system. so if package.json* is changed, then Docker will use all other layers that remain untouched, and only create the new layer were necessary.

`COPY . /usr/src/app/` is required to copy everything need (and excludes resources based on [.dockerignore](.dockerignore). So we are copying packer.json* here already.

Including an explicit copy for package.json* will mean two cached layers (at least) will be recreated, instead of just the once cached layer.

Also this OSS, and Docker image is being built by Githubs build servers and pushing to the poetapp/node on docker hub.